source/common/network/listen_socket_impl.cc

```diff
@@ -60,6 +60,35 @@ TcpListenSocket::TcpListenSocket(int fd, const Address::InstanceConstSharedPtr&
   setListenSocketOptions(options);
 }
 
+UdpListenSocket::UdpListenSocket(const Address::InstanceConstSharedPtr& address,
+                                 const Network::Socket::OptionsSharedPtr& options,
+                                 bool bind_to_port)
+    : ListenSocketImpl(address->socket(Address::SocketType::Datagram), address) {
+  RELEASE_ASSERT(fd_ != -1, "");
+
+  // TODO(htuch): This might benefit from moving to SocketOptionImpl.
+  // TODO(cmluciano): Analyze special options to be enabled/disabled for UDP
+  int on = 1;
+  int rc = setsockopt(fd_, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
+  RELEASE_ASSERT(rc != -1, fmt::format("failed to set UDP SO_REUSEADDR socket option"));
+  rc = setsockopt(fd_, SOL_SOCKET, SO_SNDBUF, &on, sizeof(on));
```

Comment:
nit: Why not add a `setOption` method in the base `Socket` class like:
```
 template <typename V>
    static int setOption(int fd, int level, int name, V* value) {
        return ::setsockopt(fd, level, name, (void*)value, sizeof(V));
    }
```

Then you can use the method as well as becomes a nice utility method for the `Socket` class.

Reply:
I like it. I think there are some notes from Harvey to do this for the TCP defaults as well.

Reply:
@htuch Per your comment [here](https://github.com/envoyproxy/envoy/blob/master/source/common/network/listen_socket_impl.cc#L45) . Should this leverage something like [SocketOptionImpl::setSocketOption](https://github.com/envoyproxy/envoy/blob/master/source/common/network/socket_option_impl.cc#L28)

Reply:
Probably.. certainly going via the OS syscall shim layer is desirable for testing, and when using socket options that are not universally supported, it's nice to have the additional machinery provided there.